torch/distributed/_functional_collectives.py

```diff
@@ -79,56 +78,72 @@ def _is_running_under_torch_deploy():
 data_ptr_to_work = dict()
 work_version = 0
 
-def _register_tensor_work(tensor, work):
-    # Note: called directly by inductor codegen currently
-    global data_ptr_to_work
-    global work_version
-    data_ptr_to_work[tensor.data_ptr()] = (work_version, work)
-    work_version += 1
+class _WaitRegistration:
+    def __init__(self, work):
+        global work_version
+        self.work = work
+        self.version = work_version
+        self.ptrs = []
+        self.cleanup_count = 0
+        work_version += 1
+
+    def _register(self, tensor):
+        global data_ptr_to_work
+        ptr = tensor.data_ptr()
+        data_ptr_to_work[ptr] = self
+        self.ptrs.append(ptr)
+        self.cleanup_count += 1
+
+    def wait(self):
+        if self.work is not None:
+            self.work.wait()
+            self.work = None
+        self.cleanup()
+
+    def decrement_live_tensor(self, ptr):
+        self.cleanup_count -= 1
+        if self.cleanup_count == 0:
+            self.cleanup()
+        else:
+            if data_ptr_to_work.get(ptr, None) == self:
+                del data_ptr_to_work[ptr]
+
+    def cleanup(self):
+        for ptr in self.ptrs:
+            if data_ptr_to_work.get(ptr, None) == self:
+                del data_ptr_to_work[ptr]
```

Comment:
when cleaning up, do we need to clean up the `self.ptrs` list too?

Reply:
`self.ptrs` is a list of ints, so it's fine to leave it as is and we don't have to worry about cleanup running multiple times as we guard against that by checking the key in the dict.